### ISSUE TITLE
Error handling enhancements

### ISSUE THREAD

**Issue by jdduh** [PSU-CSAR/vb-bagis#57]
The AOI doesn't have any snow course sites in it. An error message appears when BAGIS tries to clip the SC layer to the AOI during AOI creation. Please skip the error message if SNOTEL or SC sites do exist in the AOI so that the AOI creation can be completed without any interruption. 

![image](https://user-images.githubusercontent.com/11510708/40504092-0c031d40-5f45-11e8-981d-1d1a6db0d136.png)

**Comment by jdduh** [comment]
See #59. Please terminate the process after the first error message is dismissed.

**Comment by lbross** [comment]
Can you clarify if this happens when:
1. There is no snow course layer configured for the AOI?
2. The snow course layer exists but there are no snow course sites within the boundary of the AOI?
3. There is a snow course layer configured for the AOI but it doesn't exist in the file system?

I'm trying to recreate the issue so I can address it

**Comment by jdduh** [comment]
It’s the 2nd situation.

**Comment by lbross** [comment]
Took a while to find it, but this was a minor problem with error handling. I noticed that the AOI utilities form is configured ti return an error if the snotel and snow course sites are selected for clipping and no sites are found within the boundary. This kind of makes sense because you probably wouldn't want to reclip if the layer didn't exist in the first place, but it is inconsistent with how the AOI creation code runs. Let me know if I should amend this too.

**Comment by lbross** [comment]
@jdduh: Can you respond the question directly above? The error handling for missing snotel/snow course sites on the AOI utilities form differs from the AOI creation forms. I am going to create a new add-in today without this question answered, but we can easily handle it when we update the code to allow users to specify a buffer when re-clipping snotel/snow course.

**Comment by jdduh** [comment]
Please conform the codes to what AOI creation is doing. The ideal design is to allow uninterrupted clipping even if there are exceptions. We can prompt a summary of the exceptions (errors) when the clipping process is completed. Let me know if this response is unclear to you.

**Comment by lbross** [comment]
I understand the first part of your request and will implement as part of clipping task. Most times if there is an exception, the clipping cannot complete successfully so we need to stop. We don't currently have the infrastructure for any sort of logging or error summary in BAGIS. Is this something you would like to add next year?

**Comment by jdduh** [comment]
For this particular process, there are at most 3 tasks: clipping PRISM, SNOTEL, and SC. There tasks are (should be) independent from one another. The goal of the error handling is to prevent the errors in earlier tasks from interrupting the later tasks. The error summary would be specific to these 3 tasks, not for other BAGIS errors. Hope this make sense.

**Comment by lbross** [comment]
Are you referring only to the clipping from the AOIInfo form? Or does this also include the AOI creation forms? 

If no sites are found within the AOI boundary, should this be an error? Do you want an error message for each PRISM layer if there are multiple issues?

How do you want to display the error summary for the user? A messageBox?

Some of the error messages that pop-up originate from subroutines within the clipping process that are shared by multiple screens. And some of the subroutines are 3 or 4 layers deep. I don't know that we want to remove their pop-up messages because they may be useful for some other processes. But I can clean up the AOIInfo form some.

**Comment by jdduh** [comment]
How about this?
Two types of warning/error messages:
1. No feature within AOI, then no output layer was created. This type of error is only applicable to SNOTEL and SC featureclasses, unless there is a way to detect if the PRISM rasters cover the AOI area.
2. All other (unknown) clipping errors. The error message should be specific enough to know if the error occurred when clipping SNOTEL, SC, or PRISM (with specific layer(s)).

Please display the error messages in a messagebox. Here is a hypothetical example:

Warning: There is no SNOTEL site within the AOI. The output SNOTEL layer was not created.
Warning: There is no Snow Course site with the AOI. The output Snow Course layer was not created.
Error: Unable to clip the SNOTEL layer to the AOI. Please verify that the source layer is valid.
Error: Unable to clip the PRISM Jun layer to the AOI. Please verify that the source layer is valid.
Error: Unable to clip the PRISM Q1 layer to the AOI. Please verify that the source layer is valid.

Please try to catch the error messages that might interrupt the clipping process. Let me know if it's difficult or too troublesome. I think my bottom line for the error handling message is to know which layer was causing the error.

This is the current error message when "re-clipping" PRISM, SNOTEL, and SC layers to the Animas AOI. There is no SC site in the Animas AOI. The error message doesn't provide meaning information.
![image](https://user-images.githubusercontent.com/11510708/45114697-84270200-b102-11e8-8961-6640424e1b97.png)

**Comment by lbross** [comment]
What version of vb-bagis do you have installed? There was a problem with the error generation from the method that clips from web services but I fixed it in 3.2-b2. This is the error message that I get when clipping to Animas AOI:
![image](https://user-images.githubusercontent.com/11527966/45118239-d10fd600-b10c-11e8-8770-f5be9a91bd3b.png)

This is the current code for clipping snotel/snow course with minor differences between the two for the layer name:
```
                   If response <> 1 Then
                        Select Case response
                            Case -1 '-1: unknown error
                                MsgBox("Unknown error occurred when clipping data to AOI!")
                            Case -2 '-2: output exists
                                MsgBox("Output target layer exists in the AOI. Unable to clip new data to AOI!")
                            Case -3 '-3: missing parameters
                                MsgBox("Missing clipping parameters. Unable to clip new data to AOI!")
                            Case -4 '-4: no input shapefile
                                MsgBox("Missing the clipping shapefile. Unable to clip new data to AOI!")
                            Case 0 '0: no intersect between the input and the clip layers
                                MsgBox("No SNOTEL data exists within the AOI. Unable to clip new SNOTEL data to AOI!")
                        End Select
```
I think we could leave this mostly the same since it gives additional details but (1) add the layer name to each case statement (2) Buffer the error messages into a string rather than popping the MsgBox and then pop a MsgBox at the end with any error messages concatenated (3) Add a prefix of Warning or Error to each case statement. It sounds like Case 0 may be the only warning?

The PRISM clipping return codes are less sophisticated so we will only have a generic error message for them.

This error handling code is similar in the AOI Info form and both of the AOI creation forms. I propose that I make the changes to the AOI Info form and we can run with that for a while to see if it is what you want. If you like it, then I will copy it over to the other forms.

**Comment by lbross** [comment]
Updated error handling on AOI Utilities form and both AOI creation forms
